### PR TITLE
Update zerocopy from 0.6.6 to 0.8.8.

### PR DESCRIPTION
**Pull request by clundin25**
This PR does not optimize out a large portion of copies. Ideally it will be done in follow up PRs.

There are slight changes made to optimize panics out of the final binary.

This resolves https://github.com/chipsalliance/caliptra-sw/issues/1721

### CODE REVIEW

File: drivers/src/bounded_address.rs [hunk 1]
```diff
@@ -24,18 +24,19 @@ impl MemBounds for RomBounds {
...
+#[derive(Zeroize, TryFromBytes, Immutable)]
+pub struct BoundedAddr<T: IntoBytes + FromBytes, B: MemBounds> {
     addr: u32,
     _phantom: PhantomData<(T, B)>,
 }
-unsafe impl<T: AsBytes + FromBytes, B: MemBounds> FromBytes for BoundedAddr<T, B> {
+// Unaligned is not implemented for `u32`, so `BoundedAddr` cannot currently derive `IntoBytes`.
+// Potentially `u32` can be swapped with `U32` from the zerocopy crate, but this type does not
+// implement Zeroize.
+// We could potentially wrap `U32` with the `Zeroizing` type to resolve this.
```
Comment (jhand2) [inline drivers/src/bounded_address.rs:35]: What's the conclusion here? It seems like it's maybe working now.

Comment (clundin25) [reply drivers/src/bounded_address.rs:35]: I haven't followed up on this, but for this case I think we will not be able to use the derive feature.

I believe the orphan rule will block the workaround (Cannot implement `Zeroizing` (zeroize trait) on `U32` (Unaligned u32 struct from zerocopy).

Comment (clundin25) [reply drivers/src/bounded_address.rs]: Looked at this again with fresh eyes and cleaned it up. This type can now just derive the zerocopy traits directly.

**Review by zhalvorsen — DISMISSED** [review]
We should cherry pick this onto the 2.x branch as soon as we can because it touches so many files and tests.

**Review by jhand2 — APPROVED** [review]

**Review by jhand2 — APPROVED** [review]

File: runtime/src/get_idev_csr.rs [hunk 2]
```diff
@@ -13,44 +13,40 @@ use caliptra_error::{CaliptraError, CaliptraResult};
...
-                    // `resp.data_size` by the `IDevIDCsr::get` API.
-                    //
-                    // A valid `IDevIDCsr` cannot be larger than `MAX_CSR_SIZE`, which is the max
-                    // size of the buffer in `GetIdevCsrResp`
-                    resp.data[..resp.data_size as usize].copy_from_slice(csr);
-
-                    Ok(MailboxResp::GetIdevCsr(resp))
-                }
+        let cmd = GetIdevCsrReq::ref_from_bytes(cmd_args)
+            .map_err(|_| CaliptraError::RUNTIME_MAILBOX_INVALID_PARAMS)?;
```
Comment (swenson) [inline runtime/src/get_idev_csr.rs:24]: Before wasn't this `CaliptraError::RUNTIME_INSUFFICIENT_MEMORY`?

**Review by swenson — COMMENTED** [review]
Phew! This looked like it was a big pain to write.

Looks good; just a couple of questions

File: Cargo.toml [hunk 3]
```diff
@@ -183,7 +183,7 @@ ureg-schema = { path = "ureg/lib/schema" }
 ureg-systemrdl = { path = "ureg/lib/systemrdl" }
 wycheproof = "0.5.1"
 x509-parser = "0.15.0"
-zerocopy = "0.6.6"
+zerocopy = { version = "0.8.8", features = ["derive"] }
```
[the remark this answers is not in the document]
Comment (clundin25) [reply Cargo.toml:186]: I'll do a fast-follow up PR to bump the version, which (hopefully) will be a smaller PR.

That way I avoid dismissing approvals! 😄

File: runtime/src/get_idev_csr.rs (hunk 2 above)
Comment (clundin25) [reply runtime/src/get_idev_csr.rs:24]: This no longer is a copy, so `CaliptraError::RUNTIME_INSUFFICIENT_MEMORY` is not the failure scenario.

File: drivers/src/bounded_address.rs [hunk 4]
```diff
@@ -24,31 +24,41 @@ impl MemBounds for RomBounds {
 pub type RomAddr<T> = BoundedAddr<T, RomBounds>;
 
 #[repr(C)]
-#[derive(Zeroize)]
-pub struct BoundedAddr<T: AsBytes + FromBytes, B: MemBounds> {
-    addr: u32,
+#[derive(TryFromBytes, IntoBytes, Immutable)]
+pub struct BoundedAddr<T: IntoBytes + FromBytes, B: MemBounds> {
+    addr: Unalign<u32>,
```
[the remark this answers is not in the document]
Comment (clundin25) [reply drivers/src/bounded_address.rs:29]: Previously, the zerocopy marker traits `FromBytes` and `AsBytes` where manually implemented.

Now I have modified this type to use the zerocopy derive macros, which enforce a set of [safety conditions](https://docs.rs/zerocopy/latest/zerocopy/trait.IntoBytes.html#safety).

I have modified the `addr` field from a `u32` to `Unalign<u32>` to satisfy the safety condition _"if the type is repr(C), its fields must be [Unaligned](https://docs.rs/zerocopy/latest/zerocopy/trait.Unaligned.html)"._

**Review by swenson — APPROVED** [review]

File: Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (clundin25) [reply Cargo.toml:186]: Looks like the build currently fails on 0.8.13.

```
error[E0446]: private type `dpe_instance::_::__Zerocopy_Field_reserved` in public interface
  --> dpe/dpe/src/dpe_instance.rs:45:35
   |
45 | #[derive(IntoBytes, TryFromBytes, KnownLayout, Immutable, Zeroize)]
   |                                   ^^^^^^^^^^^
   |                                   |
   |                                   can't leak private type
   |                                   `dpe_instance::_::__Zerocopy_Field_reserved` declared as private
   |
   = note: this error originates in the derive macro `KnownLayout` (in Nightly builds, run with -Z macro-backtrace for more info)
```

The DPE crate has a private field in a public type, which is causing a zerocopy macro to trip a compiler error. I don't think that is an incorrect use case, but I need to dig deeper.

**Pull request merged**